test/integration_test/action_test.go

```diff
@@ -0,0 +1,264 @@
+
+	scaleFactor := scaleDownApps(t, ctxs)
+	logrus.Infof("scaleFactor: %v", scaleFactor)
+
+	executeOnDestination(t, func() {
+		startFailover(t, actionName, namespaces)
+		validateFailover(t, actionName, namespaces, preMigrationCtxs, true)
+	})
+}
+
+// Failover namespace without a migration on the destination
+func testFailoverWithoutMigration(t *testing.T) {
+	appKey := "mysql-nearsync"
+	instanceIDs := []string{"failover"}
+	storageClass := "px-sc"
+	actionName := "failover-action"
+	namespaces := getNamespaces(instanceIDs, appKey)
+	cleanup(t, namespaces[0], storageClass)
+
+	ctxs := scheduleAppAndWait(t, instanceIDs, appKey)
+
+	executeOnDestination(t, func() {
+		_, err := createActionCR(t, actionName, namespaces[0])
+		require.Error(t, err, "create action CR should have errored out")
+		err = schedulerDriver.WaitForRunning(ctxs[0], defaultWaitTimeout/4, defaultWaitInterval/4)
+		require.Error(t, err, "did not expect app to start")
+	})
+}
+
+// Failover multiple namespaces simultaneously
+func testFailoverForMultipleNamespaces(t *testing.T) {
+	appKey := "mysql-nearsync"
+	instanceIDs := []string{"failover-1", "failover-2"}
+	storageClass := "px-sc"
+	migrationName := "failover-migration"
+	actionName := "failover-action"
+	namespaces := getNamespaces(instanceIDs, appKey)
+	for _, namespace := range namespaces {
+		cleanup(t, namespace, storageClass)
+	}
+
+	ctxs := scheduleAppAndWait(t, instanceIDs, appKey)
+
+	startAppsOnMigration := false
+	preMigrationCtxs, ctxs, _ := triggerMigrationMultiple(
+		t, ctxs, migrationName, namespaces, true, false, startAppsOnMigration)
+
+	for idx, namespace := range namespaces {
+		validateMigrationOnSrcAndDest(
+			t, migrationName, namespace, preMigrationCtxs[idx],
+			startAppsOnMigration, uint64(4), uint64(0))
+	}
+
+	scaleFactors := scaleDownApps(t, ctxs)
+	logrus.Infof("scaleFactors: %v", scaleFactors)
+
+	executeOnDestination(t, func() {
+		startFailover(t, actionName, namespaces)
+		validateFailover(t, actionName, namespaces, preMigrationCtxs, true)
+	})
+}
+
+// Failover a namespace with multiple running applications
+func testFailoverWithMultipleApplications(t *testing.T) {
+	appKey := "mysql-nearsync"
+	additionalAppKeys := []string{"cassandra"}
+	instanceIDs := []string{"failover"}
+	storageClass := "px-sc"
+	migrationName := "failover-migration"
+	actionName := "failover-action"
+	namespaces := getNamespaces(instanceIDs, appKey)
+	cleanup(t, namespaces[0], storageClass)
+
+	ctxs := scheduleAppAndWait(t, instanceIDs, appKey)
+	addTasksAndWait(t, ctxs[0], additionalAppKeys)
+
+	startAppsOnMigration := false
+	preMigrationCtxs, ctxs, _ := triggerMigrationMultiple(
+		t, ctxs, migrationName, namespaces, true, false, startAppsOnMigration)
+
+	for idx, namespace := range namespaces {
+		validateMigrationOnSrcAndDest(
+			t, migrationName, namespace, preMigrationCtxs[idx],
+			startAppsOnMigration, uint64(4), uint64(0))
+	}
+
+	scaleFactors := scaleDownApps(t, ctxs)
```

Comment:
We should also add a test which will not scaleDownApps on the source side but a Failover action if triggered will fail.